### PR TITLE
[FLINK-10050] Support allowedLateness in CoGroupedStreams

### PR DESCRIPTION
**Pull request by EugeneYushin**
## What is the purpose of the change
[https://issues.apache.org/jira/browse/FLINK-10050](https://issues.apache.org/jira/browse/FLINK-10050)
Add 'allowedLateness' method to coGroup and join streams API.

## Brief change log

- add 'allowedLateness' for CoGroupedStreams/JoinedStreams java and scala API
- delegate calls to underlying WindowedStream (as for Trigger/Evictor scenario)

## Verifying this change

This change is a trivial rework.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

**Comment by EugeneYushin** [comment]
@aljoscha @tillrohrmann 
guys, please can you take a glance on it?

**Comment by kl0u** [comment]
Hi @EugeneYushin . This PR has no tests. Please add them.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/CoGroupedStreams.java [hunk 1]
```diff
@@ -254,7 +260,17 @@ protected WithWindow(DataStream<T1> input1,
...
+					windowAssigner, trigger, newEvictor, allowedLateness);
+		}
+
+		/**
+		 * Sets the time by which elements are allowed to be late.
+		 * @see WindowedStream#allowedLateness(Time)
+		 */
+		@PublicEvolving
+		public WithWindow<T1, T2, KEY, W> allowedLateness(Time newLateness) {
+			return new WithWindow<>(input1, input2, keySelector1, keySelector2, keyType,
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/CoGroupedStreams.java]: Check for `null`. If the user calls `allowedLateness`, he should not pass a `null` value.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/JoinedStreams.java [hunk 2]
```diff
@@ -239,7 +245,17 @@ protected WithWindow(DataStream<T1> input1,
...
-					windowAssigner, trigger, newEvictor);
+					windowAssigner, trigger, newEvictor, allowedLateness);
+		}
+
+		/**
+		 * Sets the time by which elements are allowed to be late.
+		 * @see WindowedStream#allowedLateness(Time)
+		 */
+		@PublicEvolving
+		public WithWindow<T1, T2, KEY, W> allowedLateness(Time newLateness) {
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/JoinedStreams.java]: Check for `null` argument.

Comment (EugeneYushin) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/JoinedStreams.java]: Check for null in this place breaks current logic of CoGroup/Join classes. CoGroup has no checks for nulls directly in `evictor`/`trigger` methods and validates during delegation:
https://github.com/apache/flink/blob/98412a5f7227d7694c727847727f9434bcca4e92/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/CoGroupedStreams.java#L344

I do null check there as well for `allowedLateness` field.
In the same time, Join.apply delegates to CoGroup.apply:
https://github.com/apache/flink/blob/98412a5f7227d7694c727847727f9434bcca4e92/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/JoinedStreams.java#L314). 

To be consistent, we also should add null checks for evictor and trigger. Adding null checks directly in setters/constructor breaks chain of calls in `apply` methods (for evictor/trigger/allowedLateness) and requires clumsy if-else conditions for each nullable field separately.

Both CoGroup and Join allow null for trigger/evictor (and I've added allowedLateness following the same approach) but don't pass validation during calls to `apply(...)`.

As a result of null check for `allowedLateness` inside setter, we have errors for the following scenario:
https://github.com/apache/flink/blob/8674b69964eae50cad024f2c5caf92a71bf21a09/flink-streaming-scala/src/test/scala/org/apache/flink/streaming/api/scala/CoGroupJoinITCase.scala#L151

It breaks when user doesn't specify any of trigger/evictor/allowedLateness. At the same time, these fields are optional and have defaults in WindowedStream. Unfortunately, default for allowedLateness in WindowedStream has private modificator (and I don't think it's a good practice to set default when user passed null by mistake).

Please, let me know you thoughts.

Comment (kl0u) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/JoinedStreams.java]: The `null` check should be added to check if the user-specified argument is null (the `newLateness`). Not if the class field `allowedLateness` is null. `allowedLateness` can be null, as this is the default value. This check does not break anything and they should be there to tell that if you call the method, you cannot pass a `null` as an argument.

Nobody said to put checks for the `evictor` and the `trigger`.

In addition, you should add some tests in the PR. This is not only to test if everything works, but also to guarantee that nobody in the future will break this functionality.

Comment (EugeneYushin) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/JoinedStreams.java]: @kl0u can you please take a look at https://lists.apache.org/list.html?dev@flink.apache.org
I'm in process of writing unit tests, and I can't get stable results as for now.

Do you have any thoughts on this topic?

Comment (kl0u) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/JoinedStreams.java]: I replied in the Mailing List.

**Comment by kl0u** [comment]
Thanks for updating the PR @EugeneYushin . I will have a look today.

**Comment by EugeneYushin** [comment]
@kl0u I didn't want to ping you until Travis is finished.
Thanks for quick turnaround

**Comment by EugeneYushin** [comment]
I see one build from Travis has been finished with error:
```
10:15:13.087 [ERROR] Failed to execute goal on project flink-storm-examples_2.11: Could not resolve dependencies for project org.apache.flink:flink-storm-examples_2.11:jar:1.7-SNAPSHOT: Could not transfer artifact org.apache.storm:storm-starter:jar:1.0.0 from/to central (http://repo.maven.apache.org/maven2): GET request of: org/apache/storm/storm-starter/1.0.0/storm-starter-1.0.0.jar from central failed: Connection reset -> [Help 1]
```

Looks like it's not related to proposed changes.
What is the best way to deal with such kind of failures during CI?
Is there any way to rerun PR?

**Comment by kl0u** [comment]
I would recommend to relaunch that specific build and see if it happens again.

**Comment by EugeneYushin** [comment]
Nice, CI checks have been passed. I think I have no access to Travis to relaunch builds (at least, I don't see any button there to perform such kind of operation).
Thanks for cooperation

**Comment by EugeneYushin** [comment]
@kl0u Great news!

**Pull request closed without merging**